main.go

```diff
@@ -0,0 +1,382 @@
+		resp, err = svc.ListResourceRecordSets(&req)
+		if err != nil {
+			return nil, err
+		}
+		rrsets = append(rrsets, resp.ResourceRecordSets...)
+		if *resp.IsTruncated {
+			req.StartRecordName = resp.NextRecordName
+			req.StartRecordType = resp.NextRecordType
+			req.StartRecordIdentifier = resp.NextRecordIdentifier
+		} else {
+			break
+		}
+	}
+
+	return rrsets, nil
+}
+
+// getThanosTargets is used to get only Thanos related Route53 records.
+func getThanosTargets(recordSets []*route53.ResourceRecordSet) []string {
+	thanosTargets := []string{}
+	for _, record := range recordSets {
+		if strings.Contains(*record.Name, "-grpc.") {
+			thanosTargets = append(thanosTargets, *record.Name)
+		}
+	}
+	log.Info("Returning matching Thanos Route53 records")
+	return thanosTargets
+}
+
+// createOrUpdateTargetConfigMap is used to create or update the Thanos target configmap
+func createOrUpdateTargetConfigMap(targets []string, thanosConfigMapName, thanosNamespace string, clientset *kubernetes.Clientset) error {
+
+	serviceDiscovery := fmt.Sprintf("- targets:\n  - %s", strings.Join(targets, "\n  - "))
+
+	configMapData := make(map[string]string, 0)
+	configMapData["servicediscovery.yml"] = serviceDiscovery
+	configMap := corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      thanosConfigMapName,
+			Namespace: thanosNamespace,
+		},
+		Data: configMapData,
+	}
+	ctx := context.TODO()
+	_, err := clientset.CoreV1().ConfigMaps(thanosNamespace).Get(ctx, thanosConfigMapName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if err != nil && errors.IsNotFound(err) {
+		log.Infof("Configmap %s does not exist. Creating...", thanosConfigMapName)
+		_, err = clientset.CoreV1().ConfigMaps(thanosNamespace).Create(ctx, &configMap, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Infof("Configmap %s already exists. Updating...", thanosConfigMapName)
+		_, err := clientset.CoreV1().ConfigMaps(thanosNamespace).Update(ctx, &configMap, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// rotateQueryPods is used to rotate all Thanos Query pods
+func rotateQueryPods(thanosNamespace, thanosDeploymentName string, clientset *kubernetes.Clientset) error {
+	queryPods, err := getPodsFromDeployment(thanosNamespace, thanosDeploymentName, clientset)
+	if err != nil && k8serrors.IsNotFound(err) {
+		log.Infof("No Thanos Query deployment found. Assuming that Thanos is not deployed, moving on")
+		return nil
+	} else if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	wait := 180
+
+	for _, pod := range queryPods.Items {
+		log.Infof("Deleting pod %q", pod.GetName())
+		err = deletePod(thanosNamespace, pod.GetName(), clientset)
+		if err != nil {
+			return err
+		}
+
+		queryPods, err = getPodsFromDeployment(thanosNamespace, thanosDeploymentName, clientset)
```

Comment:
This is a little weird. Your outer loop is looping through `queryPods.Items` but then you overwrite `queryPods` here. That's likely going to cause some weirdness if not panics. Looking at it closer it might actually cause the outer loop to never exit since there's always new pods for it to loop through and delete. Is the idea here you want to delete one pod, wait for the new one to be ready and then delete the next pod?

Reply:
Yeap the idea behind that is to delete the existing pods one by one but after checking that the new pod is up, running and has passed all checks. This way we have zero downtime in the central Thanos. Using the same variable name was wrong, updated it into another one.